### PR TITLE
add tls-max-version

### PR DESCRIPTION
**Pull request by pschou**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

We need to add MaxVersion for TLS to enforce configurable options.

#### Which issue(s) this PR fixes:

#81145 

#### Special notes for your reviewer:

Adds a flag for setting the maximum version.  If left off the defaults stay the same as expected by previous versions.

#### Does this PR introduce a user-facing change?

The command line flag is added for fine-tuning the versions supported with TLS sessions.

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
--tls-max-version on kubelet and kube-apiserver allow for configuring maximum TLS versions
```

### CODE REVIEW

**Review by RainbowMango — COMMENTED** [review]
> We need to add MaxVersion for TLS to enforce configurable options.

Could you please explain a little bit more about why we need to set the max version of TLS?

If the ciphers are concerning, we can set the ciphers by `--tls-cipher-suites` with preferred values or just set `--tls-min-version` to `VersionTLS13`.

**Comment by pschou** [comment]
Note in the Golang documentation:

"When we implemented TLS 1.3 in Go 1.12, we didn't make TLS 1.3 cipher suites configurable, because they are a disjoint set from the TLS 1.0–1.2"

So, in effect... With crypto/tls: TLS 1.3 ciphers are not configurable

By implementing this update one can enforce the ciphers on the command line-- if the maximum is not set, golang will add all three TLSv1.3 ciphers, effectively making the configuration of which ciphers are allowed not configurable-- it's an all or nothing setup.

This is the only way supported by Golang to fully select the ciphers.

File: pkg/kubelet/apis/config/types.go [hunk 1]
```diff
@@ -129,6 +129,9 @@ type KubeletConfiguration struct {
 	// TLSMinVersion is the minimum TLS version supported.
 	// Values are from tls package constants (https://golang.org/pkg/crypto/tls/#pkg-constants).
 	TLSMinVersion string
+	// TLSMaxVersion is the minimum TLS version supported.
```
Comment (VirrageS) [inline pkg/kubelet/apis/config/types.go:132]:
```suggestion
	// TLSMaxVersion is the maximum TLS version supported.
```

**Review by VirrageS — CHANGES_REQUESTED** [review]

**Comment by pschou** [comment]
@VirrageS thank you for that catch

File: pkg/kubelet/apis/config/types.go (hunk 1 above)
Comment (pschou) [reply pkg/kubelet/apis/config/types.go]: Thank you for noticing this, fixed!

**Comment by bart0sh** [comment]
/ok-to-test
/triage accepted
/priority important-longterm

@pschou please, add a release note, thanks.

**Comment by enj** [comment]
> Note in the Golang documentation:
> 
> "When we implemented TLS 1.3 in Go 1.12, we didn't make TLS 1.3 cipher suites configurable, because they are a disjoint set from the TLS 1.0–1.2"
> 
> So, in effect... With crypto/tls: TLS 1.3 ciphers are not configurable
> 
> By implementing this update one can enforce the ciphers on the command line-- if the maximum is not set, golang will add all three TLSv1.3 ciphers, effectively making the configuration of which ciphers are allowed not configurable-- it's an all or nothing setup.
> 
> This is the only way supported by Golang to fully select the ciphers.

The reason Go does not let you pick ciphers with TLS 1.3 is because it already picks the best ones for you (from a cryptographic standpoint all of those ciphers are secure).  I see no meaningful reason to make this change, nor would I recommend anyone set such a flag on any API server.

/hold

**Pull request closed without merging**